src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.cc

```diff
@@ -91,12 +93,14 @@ void BatchedThreadedNnet3CudaOnlinePipeline::AllocateAndInitializeData(
   d_all_log_posteriors_.Resize(max_batch_size_ * output_frames_per_chunk_,
                                trans_model_->NumPdfs(), kUndefined);
   available_channels_.resize(config_.num_channels);
+  channels_info_.reset(new std::vector<ChannelInfo>(config_.num_channels));
```

Comment:
FWIW, although `std::atomic<T>` is not movable, most containers of atomics or structs with atomics, including the `vector`, perfectly are: https://godbolt.org/z/Geda3x79a

Moving an `std::vector` boils down to moving essentially a pair of pointers (the begin and one-past-end); the memory where the _elements_ of the vector live is not moved (exotic cases involving different allocators excluded). C++ guarantees even iterator validity after move in the case of a matching allocator.

`channels_info_` thus could be a member vector more like the others, not a pointer to one. This removes one indirection (which is probably not very impactful for perf but doesn't hurt it in any case), but more importantly takes care of the element access indirection idiom like `(*vecptr)[ix]`, so that would be a readability improvement.

`resize()` won't work, as it potentially moves elements, but moving a newly constructed temporary vector will, as in the example at Godbolt's.

arm8/aarch64 also compiles fine; arm7 is not, no idea why (but is 32-bit, so no CUDA).

Reply:
I agree this is ugly (esp. the (*vec)[idx] part), but the reason for this is the `config_.CheckAndFixConfigs();` ; this function checks and may modify `config_.num_channels_`. Because of this, we cannot directly build the vector with the right size in the member init list (and if we don't, we have to resize, and move the std::atomics which are in the ChannelInfo struct).

However I think something we can do is compute the "real" num_channels_ (as a pipeline member this time) in the class constructor member init list, and then use it directly. I'll try this, because as you say, using directly the vector would be way better than using the unique_ptr to the vector

Reply:
Done

Reply:
Thanks!

The structure of the config class has been a huge pain point in productizing cudadecoder. We maintain a patched branch of Kaldi, but certainly want to integrate to vanilla code. When the construction dust settles (I just merged a large PR from @galv, and this one is also quite a change!) I'm going to refactor it for easier separation from the binary. The problem with the config is that it defines some "parameters" which are not values but rather full pathnames to other config files. A production system does not have this structure in the filesystem. In the CPU decoding feature pipelines it's also not super pretty, but at the least doable: there is always a pair of classes like [`OnlineNnet2FeaturePipelineInfo`](https://github.com/kaldi-asr/kaldi/blob/e6cc1eef1f6c55936d36e8febe9baff6c479f81b/src/online2/online-nnet2-feature-pipeline.h#L138) which reads the "master" .conf file, handles this peculiar "filesystem indirection" (which has no programmatic abstraction) and stores individual parameters in sub-structures, but may be filled in programmatically instead, and [`OnlineNnet2FeaturePipeline`](https://github.com/kaldi-asr/kaldi/blob/e6cc1eef1f6c55936d36e8febe9baff6c479f81b/src/online2/online-nnet2-feature-pipeline.h#L200-L206) constructed from this information. The cudadecoder pipelines should probably have a similar `Info` object split off following the existing pattern. I'm going to give it a go when we've committed this PR, unless you have some other large WIP part way—I don't want to run into major merge conflicts here.